### ISSUE TITLE
[rewrite:router] The previous view won't redraw while an async resolution is pending.

### ISSUE THREAD

**Issue by pygy** [lhorie/mithril.js#1268]
Using the router, you are viewing the tree corresponding to `/` and click a link to `/2` that will be resolved asynchronously.

In the interval between the `route.set()` and `resolve()` calls, the previous view becomes inert.

http://jsbin.com/fiyuciyusa/edit?js,console

```JS
m.route(document.body, '/', {
  '/':{view: function(){
    console.log("view")
  }},
  '/2':{onmatch: function(){
    console.log("Matched")
  }}
})

setTimeout(function(){m.route.set('/2')},60)

setInterval(function(){m.redraw()}, 1000)
```

Expected: `view onmatch view view view...`
Actual: `view onmatch onmatch onmatch onmatch...`
Once an async route has been requested, calls to `m.redraw()` trigger `onmatch` again rather than redrawing the previous view. Even if you prevent all but the last `resolve` to be called as describe in #1267, you end up with an unresponsive UI.

**Comment by pygy** [comment]
Maybe the title is a premature conclusion but there's a bug in here too. Do you expect each m.redraw() to run unresolved `onmatch` again and again, and freeze the UI. Imagine a network timeout. The app becomes potentialy frozen and unusable since redraws are lost.

**Comment by pygy** [comment]
A third solution (beside the mount-based one) and #1262 would be to add a `replaying` parameter to the `resolve` callback, and set it appropriately from a thin wrapper around `resolveRoute` that would be returned from the core `defineRoutes`.

However, both that and #1262 feel like clutches to me.

**Comment by pygy** [comment]
Edited the original submission to be more descriptive, less prescriptive.

**Comment by barneycarroll** [comment]
Hey @pygy I'm not sure if I'm missing something but IMO a redraw called while a RouteResolver has yet to redraw should execute the render / onbeforeupdate + view of the last route endpoint, no?

**Comment by pygy** [comment]
Yes, `m.redraw()` should affect the last fully resolved end point (that's what #1290 does).

**Comment by barneycarroll** [comment]
Ah! Too much stuff happening, and I've recently switched email clients which makes tracking all this harder than it should be. I don't know how you manage to keep focus!

**Comment by pygy** [comment]
Being the source of the mayhem sure helps :-)

**Comment by pygy** [comment]
Fixed now that #1290 has landed
